src/rbd_fuse/rbd-fuse.cc

```diff
@@ -30,212 +29,532 @@
+    if (boot_signature == 0xaa55)
+    {
+        // Read the partition information into the partitions array
+        ret = rbd_read(open_image->image, 446, 64, (char *)partitions);
+        
+        // A partition type of 0xee indicates a protective MBR for a GPT layout, so look for GPT partitions.
+        // Otherwise, any partition with a non-zero type code gets added as a file.
+        for (int i = 0; i < 4; i++)
+            if (partitions[i].part_type == 0xee)
+                get_gpt_partitions(file_descriptor);
+            else if (partitions[i].part_type != 0)
+                add_partition(i + 1, open_image->name, open_image->image, partitions[i].start_sector, partitions[i].num_sectors);
+    }
+    
+    return 0;
+}
 
-	for (im = rbd_image_data.images; im != NULL; im = im->next)
-		iter(cookie, im->image_name);
-	readdir_lock.unlock();
+// Scan the RBD image for partition tables and add any partitions as files
+int get_rbd_partitions(int file_descriptor)
+{
+    return get_mbr_partitions(file_descriptor);
 }
 
-static void count_images_cb(void *cookie, const char *image)
+int get_rbd_image(std::string image_name)
 {
-	(*((unsigned int *)cookie))++;
+    struct rbd_image_data *rbd = NULL;
+
+    if (image_name.empty())
+        return -1;
+
+    rbd_image_t temp_rbd;
+
+    // Get the next file descriptor and open the RBD
+    int file_descriptor = get_next_file_descriptor();
+    // Get the image and snapshot names from image_name
+    std::size_t snapshot_index = image_name.find('@');
+    std::string image_only_name = image_name;
+    std::string snapshot_name = "";
+    int ret = 0;
+    if (snapshot_index == std::string::npos) {
+        ret = rbd_open(ioctx, image_name.c_str(), &temp_rbd, NULL);
+    }
+    else {
+        image_only_name = image_name.substr(0, snapshot_index);
+        snapshot_name = image_name.substr(snapshot_index + 1, image_name.length() - snapshot_index - 1);
+        // Open snapshot files read-only
+        ret = rbd_open_read_only(ioctx, image_only_name.c_str(), &temp_rbd, snapshot_name.c_str());
+    }
+    if (ret < 0) {
+        simple_err("add_image_data: can't open: ", ret);
+        return ret;
+    }
+
+    // Add the image and filedescriptor to the maps
+    rbd_openimage *open_image = get_open_image(file_descriptor);
+    open_images_lock.lock();
+    open_image->name = image_name;
+    open_image->image = temp_rbd;
+    open_images_lock.unlock();
+    rbd = get_rbd_image_data(image_name);
+    image_data_lock.lock();
+    rbd->image_name = image_only_name;
+    rbd->snapshot_name = snapshot_name;
+    rbd->sector_size = imagesectorsize;
+    rbd->starting_sector = 0;
+    image_data_lock.unlock();
+    if (map_partitions) {
+        ret = get_rbd_partitions(file_descriptor);
+
+        if (ret < 0)
+            return ret;
+    }
+
+    // Stat the RBD to populate the file info
+    rbd_stat(open_image->image, &(rbd->rbd_stat.rbd_info),
```

Comment:
Seems like this would not detected externally resized images -- perhaps ```rbdfs_statfs``` and ```rbdfs_getattr``` should just quickly open the image read-only and stat it when requested?